src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java

```diff
@@ -0,0 +1,114 @@
+package org.veupathdb.service.eda.ss.model.reducer;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.gusdb.fgputil.functional.Functions;
+import org.gusdb.fgputil.functional.TreeNode;
+import org.veupathdb.service.eda.ss.model.Entity;
+import org.veupathdb.service.eda.ss.model.Study;
+import org.veupathdb.service.eda.ss.model.filter.Filter;
+import org.veupathdb.service.eda.ss.model.filter.SingleValueFilter;
+import org.veupathdb.service.eda.ss.model.reducer.ancestor.EntityIdIndexIteratorConverter;
+import org.veupathdb.service.eda.ss.model.variable.Variable;
+import org.veupathdb.service.eda.ss.model.variable.VariableValueIdPair;
+import org.veupathdb.service.eda.ss.model.variable.VariableWithValues;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Constructs data flow trees for tabular subsetting requests. These trees represent the flow of data for a particular
+ * request. The tree will be rooted with the entity for which we are outputting one or more variables.
+ */
+public class DataFlowTreeFactory {
+  private static final Logger LOG = LogManager.getLogger(DataFlowTreeFactory.class);
+
+  private final BinaryValuesStreamer binaryValuesStreamer;
+  private final EntityIdIndexIteratorConverter idIndexIteratorConverter;
+
+  public DataFlowTreeFactory(BinaryValuesStreamer binaryValuesStreamer,
+                             EntityIdIndexIteratorConverter idIndexIteratorConverter) {
+    this.binaryValuesStreamer = binaryValuesStreamer;
+    this.idIndexIteratorConverter = idIndexIteratorConverter;
+  }
+
+  /**
+   * Creates a data-flow tree from the pruned entity tree. The constructed tree will be rooted with the outputEntity
+   * passed in.
+   * @param prunedEntityTree Entity tree pruned to only active nodes based on output and filters.
+   * @param outputEntity Entity for which variables are output.
+   * @param filters Filters to apply to entity stream.
+   * @param outputVariables Variables to return for each output entity.
+   * @param study Study for which
+   * @return
+   */
+  public DataFlowMapReduceTree create(TreeNode<Entity> prunedEntityTree,
+                                      Entity outputEntity,
+                                      List<Filter> filters,
+                                      List<Variable> outputVariables,
+                                      Study study) {
+    if (!outputVariables.stream().allMatch(var -> var.getEntity().equals(outputEntity))) {
+      throw new IllegalArgumentException("All output variables must be associated with output entity.");
+    }
+    try {
+      LOG.info("Creating map-reduce data-flow tree rooted with output entity: " + outputEntity.getId());
+      final TreeNode<Entity> parent = prunedEntityTree.findFirst(entity -> entity.equals(outputEntity));
```

Comment:
i dont know i could state what is going on in this method, so i'd suggest inline comments.

i was wondering why we have to do a find to locate our node in the tree.  could somebody at a higher level be sure we always get the relevant subtree?  just a thought.

Reply:
We could. I'd just need to update the FgpUtil `TreeNode` to store a reference to its parent.